src/wallet/rpcdump.cpp

```diff
@@ -71,6 +71,63 @@ std::string DecodeDumpString(const std::string &str) {
     return ret.str();
 }
 
+bool GetWalletAddressesForKey(CWallet * const pwallet, const CKeyID &keyid, std::string &strAddr, std::string &strLabel)
+{
+    bool fLabelFound = false;
+    CKey key;
+    pwallet->GetKey(keyid, key);
+    if (!key.IsCompressed()) {
+        strAddr = EncodeDumpString(pwallet->mapAddressBook[keyid].name);
+        if (pwallet->mapAddressBook.count(keyid)) {
+            strLabel = EncodeDumpString(pwallet->mapAddressBook[keyid].name);
+            fLabelFound = true;
+        }
+    } else {
+        CTxDestination p2wpkhAddr;
+        CTxDestination p2shAddr;
+        CScript p2wpkhScr = GetScriptForDestination(WitnessV0KeyHash(keyid));
+        CScript p2shScr = GetScriptForDestination(CScriptID(p2wpkhScr));
+        ExtractDestination(p2wpkhScr, p2wpkhAddr);
+        ExtractDestination(p2shScr, p2shAddr);
```

Comment:
You should iterate the result of 'GetAllDestinationsForKey(key)` here instead of replicating the logic to compute the destinations.

Reply:
That... makes a lot more sense :)